src/invidious/yt_backend/youtube_api.cr

```diff
@@ -404,38 +404,33 @@ module YoutubeAPI
     url = "#{endpoint}?key=#{client_config.api_key}"
 
     headers = HTTP::Headers{
-      "Content-Type" => "application/json; charset=UTF-8",
+      "Content-Type"    => "application/json; charset=UTF-8",
+      "Accept-Encoding" => "gzip, deflate",
     }
 
-    # The normal HTTP client automatically applies accept-encoding: gzip,
-    # and decompresses. However, explicitly applying it will remove this functionality.
-    #
-    # https://github.com/crystal-lang/crystal/issues/11252#issuecomment-929594741
-    {% unless flag?(:disable_quic) %}
-      if CONFIG.use_quic
-        headers["Accept-Encoding"] = "gzip"
-      end
-    {% end %}
-
     # Logging
     LOGGER.debug("YoutubeAPI: Using endpoint: \"#{endpoint}\"")
     LOGGER.trace("YoutubeAPI: ClientConfig: #{client_config}")
     LOGGER.trace("YoutubeAPI: POST data: #{data}")
 
     # Send the POST request
-    if client_config.proxy_region
-      response = YT_POOL.client(
-        client_config.proxy_region,
+    if {{ !flag?(:disable_quic) }} && CONFIG.use_quic
+      # Using QUIC client
+      response = YT_POOL.client(client_config.proxy_region,
         &.post(url, headers: headers, body: data.to_json)
       )
+      body = response.body
     else
-      response = YT_POOL.client &.post(
-        url, headers: headers, body: data.to_json
-      )
+      # Using HTTP client
+      body = YT_POOL.client(client_config.proxy_region) do |client|
+        client.post(url, headers: headers, body: data.to_json) do |response|
+          self._decompress(response.body_io, response.headers["Content-Encoding"]?)
+        end
+      end
```

Comment:
if you saved the response to a variable instead of using a block, it would match exactly the quic side of the condition. It seems the only difference is how the code handles the response body. The request piece could be pulled out of the if and de-duplicated.

Reply:
The problem is that the instance variable `body_io` isn't accessible outside of a block (I did what you say at the beginning, but It didn't work, so I reached to the `#crystal-lang` IRC channel for advice, and that's the response I got. Check the gitter.im link in the first post for the relevant conversation log)